[MRG+1] [DOC] Adding GMM to plot_cluster_comparison.py

gte620v opened the pull request:
## Changes
- Added GMM to plot_cluster_comparison.py
- Changed number of clusters in all algorithms to three since one of the datasets clearly has three clusters. As is, it looks like several of the clustering algorithms are unable to identify clearly distinct clusters.
## Output

Once the docs are built, the example plot looks like this:
![image](https://cloud.githubusercontent.com/assets/2158394/12878749/c60e4e62-cdf7-11e5-9281-c7ba7edac38c.png)

gte620v committed:
Adding GMM to plot_cluster_comparison.py and changing number of components in all algos to 3.

gte620v committed:
Adding GMM to plot_cluster_comparison.py and changing number of components in all algos to 3.

GaelVaroquaux commented:
Looks great in general.

The idea of setting the number of clusters to 2 was to explore the failure mode of clustering. But I have no strong feeling in this direction.

If we keep it at two, we should at least explain in the docstring why we set it this way. WDYT?

gte620v commented:
@GaelVaroquaux I don't feel super strongly about it either.  It just seems that all of these algos have at least one parameter that needs to be specified and by intentionally mis-tuning the parameter for the algos that require `n_clusters` as an input and not mis-tuning the parameters for the other algos, we somewhat misrepresent the relative performance. 

Another possibility is to set the number of clusters higher than three to see the failure mode in that direction.  Or, we could just make two or three plots: with n=2, n=3, and n=4 (this might be the best option).


But yes, either way, I think an explanation would be good.  Something like

>The results could be improved by tweaking the parameters for each clustering strategy, for instance setting the number of clusters for the methods that needs this parameter specified.

to 

>The results could be improved by tweaking the parameters for each clustering strategy, for instance setting the number of clusters for the methods that needs this parameter specified.  In these examples the number of clusters for the algorithms that require `n_clusters` as a parameter has been set to two, which is clearly a mismatch with the datasets in the bottom two rows. 

Let me know what you think and I am happy to modify the PR.

amueller commented:
well with 3 the first two plots are a bit more weird. And having three times as many plots seems not very helpful to me.

I feel that having the 2nd or 3rd of these datasets would be more interesting: http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_assumptions.html

For the current datasets, there is not that much of a difference between k-means and GMMs.

gte620v commented:
@amueller Good suggestion.  

Adding those two datasets definitely paints a broader picture of how each algo performs.

Let me know what you think and I can push those changes.

![image](https://cloud.githubusercontent.com/assets/2158394/12986800/295c4bbe-d0c7-11e5-9967-3983947aa98d.png)

jakevdp commented:
+1 on adding additional datasets

gte620v committed:
adding two datasets to clustering comparision example

gte620v committed:
adding two datasets to clustering comparision example

gte620v commented:
@jakevdp @amueller @GaelVaroquaux 

I pushed the changes. The built output looks like this:

![image](https://cloud.githubusercontent.com/assets/2158394/13345382/62bfaf0c-dc2c-11e5-848b-736d84850da6.png)

tguillemot commented:
The new `GaussianMixture` class is merged #6666.
Could you make the changes ?

gte620v commented:
@tguillemot: I'll take a look.

gte620v committed:
Merge branch 'gmm_example' of github.com:gte620v/scikit-learn into gmm_example

gte620v committed:
Merge remote-tracking branch 'upstream/master' into gmm_example

gte620v committed:
GMM example using GaussianMixture

gte620v commented:
@tguillemot i made the modification.  It should be ready to merge.

gte620v commented:
@jakevdp @amueller @GaelVaroquaux let me know if there is anything I else I need to do to make this mergeable.

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -42,21 +42,34 @@
-
-datasets = [noisy_circles, noisy_moons, blobs, no_structure]
+# noisy_circles, noisy_moons, blobs, no_structure, varied
+datasets = [noisy_circles, noisy_moons, blobs, no_structure, varied,aniso]
```
Add a space before aniso

tguillemot replied:
It's not a big deal but I think it's better to put no_structure at the end.
It's just to easily compare the result between the gaussian datas.

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -42,21 +42,34 @@
 plot_num = 1
-
-datasets = [noisy_circles, noisy_moons, blobs, no_structure]
+# noisy_circles, noisy_moons, blobs, no_structure, varied
```
Remove this comment

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -42,21 +42,34 @@
+X, y = datasets.make_blobs(n_samples=n_samples, random_state=random_state)
+transformation = [[0.6, -0.6], [-0.4, 0.8]]
+X_aniso = np.dot(X, transformation)
+aniso = (X_aniso,y)
```
Put a space before y.

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -42,21 +42,34 @@
+varied = datasets.make_blobs(n_samples=n_samples,
+                                cluster_std=[1.0, 2.5, 0.5],
+                                random_state=random_state)
+
```
Align cluster_std and random_state with n_samples.
Use flake8 to check those things.

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -27,7 +27,7 @@
+from sklearn import cluster, datasets, mixture
 from sklearn.neighbors import kneighbors_graph
 from sklearn.preprocessing import StandardScaler
 
```
I know it is not something you put but move `print(__doc__)` after the import.

gte620v replied:
I am happy to do this, but all of the other `plot_*.py` examples have this line before the imports.  Maybe it is so you can see progress when you `make` the docs?  

Any other opinions on whether this should be removed?

tguillemot replied:
I've said that because I add a warning with my sublime text extension but maybe I've configured it to be hard :).
Flake8 and pep8 don't say anything indeed, sorry for that.

gte620v replied:
no problem!

tguillemot commented:
@gte620v Thanks for the modification.
I know it is not you who did that but there is some `UserWarning` during the execution. 
Could you try to remove them ?

It's really a good idea to add GMM to that comparison. Thx

gte620v committed:
fixing lint errors; changing order of datasets in the columns so that no_structure is at the end.

gte620v commented:
@tguillemot Thanks for the review!

I made the lint fixes and reorganized the order of the datasets.  I am unsure about removing the `print(__doc__)` line as all the other examples have that line.

The `UserWarning` errors are due to some change in how random integers should be created.  For example:

```
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:1279: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, init_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:630: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, init_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:630: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, init_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:630: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, init_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:1328: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, self.batch_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:1328: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, self.batch_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:1328: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
  0, n_samples - 1, self.batch_size)
/Users/bob/code/scikit-learn/sklearn/cluster/k_means_.py:1328: DeprecationWarning: This function is deprecated. Please call randint(0, 1499 + 1) instead
```

See
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/k_means_.py#L629-L630
and
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/k_means_.py#L1327-L1328

I can fix these but it would significantly convolute this PR if I start mucking around with the core k means code.  I think it would be best to leave that to another PR.  Let me know if you see another way of doing 
it.

tguillemot commented:
Yes, It's caused by a deprecation function in numpy.
I've sent #6712 to correct the problem with rand_int.
But there is some other warning cause by the code : 
>spectral_embedding_.py:229: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.
hierarchical.py:194: UserWarning: the number of connected components of the connectivity matrix is 2 > 1. Completing it to avoid stopping the tree early.
hierarchical.py:419: UserWarning: the number of connected components of the connectivity matrix is 2 > 1. Completing it to avoid stopping the tree early.

gte620v commented:
I am not sure what is causing those warnings.  They also appear in this example in `master` and seems to come up in several issues/PRs.  e.g.:

https://github.com/scikit-learn/scikit-learn/pull/4313
https://github.com/scikit-learn/scikit-learn/issues/5327

From `master`:

```
scikit-learn/sklearn/manifold/spectral_embedding_.py:229: UserWarning: Graph is not fully connected, spectral embedding may not work as expected.
  warnings.warn("Graph is not fully connected, spectral embedding"
scikit-learn/sklearn/cluster/hierarchical.py:194: UserWarning: the number of connected components of the connectivity matrix is 2 > 1. Completing it to avoid stopping the tree early.
  connectivity, n_components = _fix_connectivity(X, connectivity)
scikit-learn/sklearn/cluster/hierarchical.py:419: UserWarning: the number of connected components of the connectivity matrix is 2 > 1. Completing it to avoid stopping the tree early.
  connectivity, n_components = _fix_connectivity(X, connectivity)
```

I am afraid, I do not know enough about the inner workings of the algos to fix this.  Let me know if you have suggestions.

tguillemot commented:
Thanks you to have tried to solve that. 
There is the same problem in `master` indeed and I think it could be cool to solve it by the same time.
I will investigate that problem next week.

gte620v commented:
Great; thanks.

tguillemot commented:
After investigation, this warning came from the `kneighbors_graph` for the datasets : `noisy_moons` and `blobs`. It warns because `kneighbors_graph` (used by `cluster.AgglomerativeClustering`) defines a connectivity graph from the 10 nearest neighbors and creates 2 or 3 connected components rather than 1 (as needed by the algorithm).

The warning is raised because, we need a unique connected component and it has to create some connection to complete it. The only way to solve that is to increase the number of nearest neighbors but it will increase the computation time for no better result.

Consequently, I suggest to catch this warning with `ignore_warning` by specifying `category=UserWarning`. Are you OK with that ?

For those datasets, it's difficult

amueller commented:
yeah that sounds good

amueller commented:
I'm not a 100% on changing all the cluster numbers to 3. We could set the parameters of the other algorithms such that they produce the "correct" number of clusters? Having the correct number of clusters for some of the datasets but not for others also seems weird.

We could have two examples, one for "well set parameters" and one for "misspecified parameters"? 

Having #4301 would be sweet ^^

gte620v commented:
Setting the number of clusters for `noisy_circles` and `noisy_moons` to 2 makes sense to me.  What about for `no_structure`: 1, 2, or 3?  #4301 is super cool.

@tguillemot Nice find.  I'll work on this in the morning.

gte620v committed:
adding warning supression.

gte620v committed:
fixing warning supression.

gte620v committed:
hand-tuned cluster parameters

gte620v commented:
- I modified the doc string to clarify what the example is demonstrating.
- Hand-tuned the cluster parameters to represent a best-effort clustering result for each dataset.  for some of the algo-datasets, I couldn't find a parameter set that had a number of clusters that matched the true number of clusters (e.g. affinity propagation and mean shift).  
- For the no_structure dataset, I just kept the parameters from the blobs dataset and added an explanation to the doc string saying as much.

![image](https://cloud.githubusercontent.com/assets/2158394/14995212/93bb694e-1141-11e6-9aaf-7a3082addf8f.png)

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,168 @@
 np.random.seed(0)
 
+colors = np.array([x for x in 'bgrcmykbgrcmykbgrcmykbgrcmyk'])
+colors = np.hstack([colors] * 20)
```
I don't like this.
Create color as `['r', 'g', ... 'b']` and then use the modulo during the plot.

tguillemot replied:
Ok maybe not the modulo but I think there is a better way to plot those things.

amueller replied:
cycler?

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,168 @@
-        two_means, affinity_propagation, ms, spectral, ward, average_linkage,
-        dbscan, birch]
+    clustering_algorithms = [two_means, affinity_propagation, ms, spectral,
+                             ward, average_linkage, dbscan, birch, gmm]
```
I think it's better to have a dictionary for that : `{name, Class(...)}`.

gte620v committed:
moved list of algo names; cleaning up color cycling

gte620v commented:
@tguillemot thanks for the comments.  I fixed both issues.

gte620v commented:
@amueller anything else I need to do to get this merged?

tguillemot commented:
@amueller Do you think you will have time to review that during the sprint ?

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,173 @@
-        plt.xlim(-2, 2)
-        plt.ylim(-2, 2)
+
+        colors = np.array(list(islice(cycle('bgrcmyk'), max(y_pred) + 1)))
```
I have a bug on my computer with `islice`.
The correction of that is : `int(max(y_pred)  + 1)`

amueller commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,173 @@
+from sklearn import cluster, datasets, mixture
 from sklearn.neighbors import kneighbors_graph
 from sklearn.preprocessing import StandardScaler
+from itertools import cycle, islice
```
islice is python2, right?

tguillemot replied:
I've python 3.5 and it's working with me but I will have check that.

gte620v replied:
This works in python 3.5

gte620v committed:
fixing islice stop to be an int

gte620v commented:
Thanks for looking!

@tguillemot I made that islice `stop` argument an int.

@amueller islice seems to work on 2.7, 3.4, 3.5.
https://docs.python.org/2.7/library/itertools.html#itertools.islice
https://docs.python.org/3.4/library/itertools.html#itertools.islice
https://docs.python.org/3.5/library/itertools.html#itertools.islice

tguillemot commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,174 @@
+for i_dataset, (dataset, params) in enumerate(datasets):
+    # update parameters with dataset-specific values
+    defaults = default_base.copy()
+    defaults.update(params)
```
Rather than creating a new defaults var you can use `estimator.set_params` where estimator is your estimator.

gte620v replied:
In thinking about this, I can see how we might move the estimators out of the loop and then update the parameters with `set_params`, but I don't see how we get rid of the defaults variable without replacing it with something more verbose.  

Since not all keywords need to be changed for each dataset, I would have to have another bit of code that checks to see if a keyword that we need is in `params` or not for that dataset.  If it isn't, I'd have to revert back to the `default_base` values.  I think updating the keyword dict `defaults`, as we have, is the simplest way to have parameters that get changed by dataset-estimator.  

The other option is to have the `params` have both the `default_base` and dataset-specific values enumerated for each dataset, but with that it is much harder to see where the dataset parameter excursions are.

Let me know if I am missing your point.

gte620v replied:
@tguillemot any thoughts?

jnothman replied:
I think it is ugly naming it `defaults` below when you mean `params`. Either use `params.get('key', default_val)` or call it `params` within the loop.

tguillemot commented:
LGTM after you correct my comments.

gte620v commented:
@tguillemot I didn't follow your suggestion above.  Can you provide clarification?  Otherwise I think it is gtg.   If we don't want to merge, that is cool too; we can close the PR.

tguillemot commented:
@gte620v Sorry for the late answer. What I said was not important but your examples are.

LGTM.

@raghavrv Can you review this PR and merge it ?

jnothman commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,174 @@
-create many clusters. Thus in this example its two parameters
-(damping and per-point preference) were set to mitigate this
-behavior.
+but still in 2D. With the expection of the last dataset,
```
*exception

gte620v replied:
Fixed

TomDLT commented on examples/cluster/plot_cluster_comparison.py:
```diff
@@ -3,116 +3,174 @@
-        plt.xlim(-2, 2)
-        plt.ylim(-2, 2)
+
+        colors = np.array(list(islice(cycle('bgrcmyk'),
```
Use `['navy', 'yellowgreen', 'gold', 'cornflowerblue']` to be color-blind compatible.

gte620v replied:
That looks like this:

![image](https://cloud.githubusercontent.com/assets/2158394/23858768/f1efcaae-07d7-11e7-8941-92556258f674.png)


cool?

gte620v replied:
The latest commit makes the above plot.

gte620v committed:
change default to params, make plot color-blind compatible, fix spelling error

gte620v commented:
Here is [another option](https://gist.github.com/thriveth/8560036) for color-blind-friendly colors:

![image](https://cloud.githubusercontent.com/assets/2158394/23859644/aecd1260-07da-11e7-85d8-98a9380a257f.png)


@TomDLT your color cycle seems like it would be problematic for blue–yellow color blindness (this is what is currently in the PR).  what do you think?

![image](https://cloud.githubusercontent.com/assets/2158394/23859690/da327cd8-07da-11e7-9179-c9fd9c3fceab.png)

Here is the original for reference:

![image](https://cloud.githubusercontent.com/assets/2158394/23859752/1ffc953c-07db-11e7-9f11-07147068490a.png)

TomDLT commented:
> @TomDLT your color cycle seems like it would be problematic for blue–yellow color blindness (this is what is currently in the PR). what do you think?

I have no strong feeling about it. The colors I suggested are used in other examples in scikit-learn, yet if you think this one is better I am fine with it.

gte620v committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into gmm_example

gte620v committed:
new color palette that is more color-blind friendly

gte620v commented:
@TomDLT Ok, great.  I picked the first colormap for the latest commit.

I think all is good to merge now.

gte620v commented:
@jnothman @TomDLT @tguillemot 
Are any other changes needed?

jmschrei commented:
This looks good to me, but I'm going to let one of the people who have been involved in the discussion for longer be the one to merge it. Thanks for the contribution and patience!

TomDLT commented:
Thanks for the patience @gte620v !

gte620v commented:
Np. Thanks!!

tguillemot commented:
Thanks @gte620v
